### PR TITLE
Update Mqtt direct method topics for E4K

### PR DESCRIPTION
**Pull request by brycewang-microsoft**

### CODE REVIEW

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java [hunk 1]
```diff
@@ -27,14 +27,14 @@ class MqttDirectMethod extends Mqtt
...
+    private static final String POST = "POST";
+    private static final String RES = "res";
+    private static final String REQ_ID = "$rid=";
 
-    //Placement for $iothub/methods/POST/{method name}/?$rid={request id}
-    private static final int METHOD_TOKEN = 3;
-    private static final int REQID_TOKEN = 4;
+    //Placement for $iothub/methods/{clientId}/POST/{methodName}/?$rid={request id}
+    private static final int METHOD_TOKEN = 4;
+    private static final int REQID_TOKEN = 5;
```
Comment (brycewang-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:37]: Now we have a new field "/{clientId}/" inserted between "$iothub/methods" and "POST", so move these 2 parts backward one place.

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java [hunk 2]
```diff
@@ -27,14 +27,14 @@ class MqttDirectMethod extends Mqtt
     private static final String BACKSLASH = "/";
     private static final String QUESTION = "?";
 
-    private static final String METHOD = "$iothub/methods/";
-    private static final String POST = METHOD + "POST";
-    private static final String RES = METHOD + "res";
-    private static final String REQ_ID = QUESTION + "$rid=";
+    private static final String METHOD = "$iothub/methods";
```
Comment (timtay-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:30]: Same comment about the number of constants as I made in #1713. We don't need a constant for every character/word here.

**Review by timtay-microsoft — CHANGES_REQUESTED** [review]
You are changing the topics regardless of if the E4K flag is set

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java [hunk 3]
```diff
@@ -250,7 +250,8 @@ else if (proxySettings.getProxy().type() == Proxy.Type.HTTP)
             deviceId,
             connectOptions,
             unacknowledgedSentMessages,
-            receivedMessages);
+            receivedMessages,
+            this.config.isConnectingToMqttGateway());
```
Comment (brycewang-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttIotHubConnection.java:254]: pass the flag `isConnectingToMqttGateway` into ctor of MqttDirectMethod

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java [hunk 4]
```diff
@@ -133,42 +145,43 @@ public IotHubTransportMessage receive()
...
-                            }
+                            message = new IotHubTransportMessage(new byte[0], MessageType.DEVICE_METHODS);
+                        }
+
+                        message.setDeviceOperationType(DeviceOperations.DEVICE_OPERATION_UNKNOWN);
+                        message.setQualityOfService(mqttMessage.getQos());
+
+                        String methodName = this.isConnectingToMqttGateway ?
+                                topicParser.getMethodName(4) :
+                                topicParser.getMethodName(3);
```
Comment (brycewang-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:170]: the value of methodName can be obtained from the different fields, based on if we are connecting to E4K or not: 4 if yes ($iothub/methods/testDevice/POST/**testMethod**/?$rid=10), 3 if no ($iothub/methods/POST/**testMethod**/?$rid=10).

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java [hunk 5]
```diff
@@ -133,42 +145,43 @@ public IotHubTransportMessage receive()
...
+                        message.setMethodName(methodName);
+
+                        String reqId = this.isConnectingToMqttGateway ?
+                                topicParser.getRequestId(5) :
+                                topicParser.getRequestId(4);
```
Comment (brycewang-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:175]: Similar here, getting requestId from the field 5 if connecting to E4K, or 4 if not.

($iothub/methods/testDevice/POST/testMethod/?$**rid=10**)
($iothub/methods/POST/testMethod/?**$rid=10**)

**Comment by brycewang-microsoft** [comment]
> You are changing the topics regardless of if the E4K flag is set

corrected

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java [hunk 6]
```diff
@@ -21,31 +21,44 @@ class MqttDirectMethod extends Mqtt
...
+            // $iothub/methods/POST/#
+            this.subscribeTopic = POST_HUB + "/#";
+
+            // $iothub/methods/res
+            this.responseTopic = RES_HUB;
+        }
+        else
+        {
+            // $iothub/methods/{clientId}/POST/#
+            this.subscribeTopic = String.format(POST_E4K, deviceId) + "/#";
```
Comment (timtay-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:57]: Should ```deviceId``` by ```deviceId/modules/moduleId``` for module cases here?

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java (hunk 4 above)
Comment (timtay-microsoft) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:170]: Can you leave this as a comment in code for future reference?

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java (hunk 6 above)
Comment (timtay-microsoft) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:57]: Actually, it looks like we weren't doing that before, either. Odd. I'll take a look at why our existing SDK doesn't seem to subscribe to module-specific methods here.

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java (hunk 6 above)
Comment (drwill-ms) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:57]: clientId is just `{deviceId}/{moduleId}`.

Comment (brycewang-microsoft) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:57]: > clientId is just `{deviceId}/{moduleId}`.

I just realized that MqttMessaging in our existing SDK is using "deviceId/modules/moduleId" as "[clientId](https://github.com/Azure/azure-iot-sdk-java/blob/main/iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttMessaging.java#L52)" in topics, so we may also need to replace this with "deviceId/moduleId" for E4K topics, right?

Comment (timtay-microsoft) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/MqttDirectMethod.java:57]: Yes. And when subscribing to direct methods, we will need to include this deviceId/moduleId as the clientId. In our existing SDK, it is fine that the subscription topic is the same for devices and modules

**Comment by brycewang-microsoft** [comment]
We are working on patching the v2 Java SDK to use Mqtt 5 for E4K, so closing this.

**Pull request closed without merging**